### PR TITLE
[WIP] Named instances

### PR DESCRIPTION
**Pull request by glibsm**
Consolidating all PRs into one WIP, since it's really only useful to see the atomic addition.

Needs a rebase, but comment away on the approach.

### CODE REVIEW

File: dig.go [hunk 1]
```diff
@@ -206,10 +204,33 @@ func (c *Container) isAcyclic(n node) error {
...
 }
 
-// Retrieve a type from the container
-func (c *Container) get(t reflect.Type) (reflect.Value, error) {
-	if v, ok := c.cache[t]; ok {
-		return v, nil
+type keyOpt func(*key)
+
+// Specifies if absence of the object in the graph is ok on retrieval.
+func optional(opt bool) keyOpt {
```
Comment (abhinav) [inline dig.go]: Optional: This pattern is probably not required here. Seems to me that `key`/`edge` expresses all the requirements for the query for a node. So you could probably just write `get(key) (reflect.Value, error)`.

File: dig_test.go [hunk 2]
```diff
@@ -462,6 +462,35 @@ func TestEndToEndSuccess(t *testing.T) {
 			require.NotNil(t, a, "*A should be part of the container through Ret2->Ret1")
 		}))
 	})
+
+	t.Run("named instances", func(t *testing.T) {
```
Comment (alsamylkin) [inline dig_test.go]: More tests are in order? I'll think about circular dependency resolution(named instances could have a different dependency graphs that are valid, but could look like a loop if you drop names), what should we return if there is only a named instance(or two) for the unnamed type and reverse.

File: dig.go [hunk 3]
```diff
@@ -206,10 +204,33 @@ func (c *Container) isAcyclic(n node) error {
...
+	return func(k *key) {
+		k.optional = opt
+	}
+}
+
+// Represents an edge in the graph.
+type key struct {
+	t        reflect.Type
+	optional bool
```
Comment (glibsm) [inline dig.go]: no, it's a property of the constructor

File: dig_test.go (hunk 2 above)
Comment (glibsm) [inline dig_test.go]: was tagged as WIP for a reason

File: dig.go [hunk 4]
```diff
@@ -37,14 +37,12 @@ const (
 // effectively singletons.
 type Container struct {
 	nodes map[reflect.Type]node
-	cache map[reflect.Type]reflect.Value
```
Comment (glibsm) [inline dig.go]: they do, because `reflect.Type` alone doesn't uniquely identify an object anymore

**Comment by glibsm** [comment]
Closed in favor of #109

**Pull request closed without merging**